### PR TITLE
Dash parameter

### PR DESCRIPTION
**Pull request by PhillypHenning**
## Summary of changes
- Added ENV check and load of critical run variables for convert-schema script
- Implemented KEYS_LIST TODO
- Added Selection-List parameter

### CODE REVIEW

File: scripts/bitops-config/convert-schema.sh [hunk 1]
```diff
@@ -22,36 +43,29 @@ function build_keys_list(){
...
     fi
   done <<< "$keys"
 }
 
-# TODO
-# KEYS_LIST="$(build_keys_list)"
-build_keys_list "$ROOT_KEY" "$ROOT_KEY_SCHEMA"
-
-
+KEYS_LIST="$(build_keys_list $ROOT_KEY)"
```
[the remark this answers is not in the document]
Comment (mickmcgrath13) [reply scripts/bitops-config/convert-schema.sh:64]: @PhillypHenning can you add this?

File: scripts/bitops-config/convert-schema.sh [hunk 2]
```diff
@@ -63,37 +74,46 @@ while IFS= read -r value; do
...
+
   type="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.type")"
   parameter="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.parameter")"
   terminal="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.terminal")"
   required="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.required")"
   export_env="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.export_env")"
   default="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.default")"
+  dash_type="$($SCRIPTS_DIR/bitops-config/get.sh $SCHEMA_FILE "${full_value_path_schema}.dash_type")"
 
+  # Default to single dash
```
Comment (mickmcgrath13) [inline scripts/bitops-config/convert-schema.sh:90]:
```
# Default to double dash
```

File: scripts/bitops-config/convert.sh [hunk 3]
```diff
@@ -1,20 +1,29 @@
...
-export BITOPS_DIR="/opt/bitops"
-export SCRIPTS_DIR="$BITOPS_DIR/scripts"
+if [ -z "$BITOPS_DIR" ];then
+  echo "Using default BitOps Directory"
+  export BITOPS_DIR="/opt/bitops"
+fi
+
+if [ -z "$SCRIPTS_DIR" ];then
+  echo "Using default BitOps Script Directory"
+  export SCRIPTS_DIR="/opt/bitops"
```
Comment (mickmcgrath13) [inline scripts/bitops-config/convert.sh:11]:
```
export SCRIPTS_DIR="/opt/bitops/scripts"
```

File: scripts/helm/bitops.schema.yaml [hunk 4]
```diff
@@ -9,11 +9,13 @@ helm:
           parameter: namespace
           export_env: NAMESPACE
           required: true
+          dash_type: '--'
```
Comment (mickmcgrath13) [inline scripts/helm/bitops.schema.yaml:12]: remove this as default is `--`

File: scripts/helm/bitops.schema.yaml [hunk 5]
```diff
@@ -9,11 +9,13 @@ helm:
...
         timeout:
           type: string
           parameter: timeout
           export_env: TIMEOUT
           default: "500s"
+          dash_type: "---"
```
Comment (mickmcgrath13) [inline scripts/helm/bitops.schema.yaml:18]: remove this

File: scripts/helm/bitops.schema.yaml [hunk 6]
```diff
@@ -25,12 +27,16 @@ helm:
         atomic:
           type: boolean
           parameter: atomic
+          dash_type: "-"
```
Comment (mickmcgrath13) [inline scripts/helm/bitops.schema.yaml:30]: remove this as it needs to be `--` (i.e. the default)

File: scripts/helm/bitops.schema.yaml [hunk 7]
```diff
@@ -25,12 +27,16 @@ helm:
...
         force:
           type: boolean
           parameter: force
         dry-run:
           type: boolean
           parameter: dry-run
+        backend-config:
```
Comment (mickmcgrath13) [inline scripts/helm/bitops.schema.yaml:37]: `backend-config` is for terraform, not helm

Comment (mickmcgrath13) [reply scripts/helm/bitops.schema.yaml:37]: it also needs `dash_type: "-"`

File: scripts/terraform/bitops.schema.yaml [hunk 8]
```diff
@@ -12,6 +12,13 @@ terraform:
           type: string
           parameter: target
           export_env: TF_TARGET
+        backend-config:
+          type: selection-list
+          parameter: cli-flag-type
+          selection-list:
```
Comment (mickmcgrath13) [inline scripts/terraform/bitops.schema.yaml:18]: shouldn't this be `dash_type: "-"` instead of `selection-list: ["-","--"]` ?

**Review by mickmcgrath13 — CHANGES_REQUESTED** [review]

File: scripts/helm/bitops.schema.yaml (hunk 4 above)
Comment (PhillypHenning) [reply scripts/helm/bitops.schema.yaml:12]: Fixed.

File: scripts/helm/bitops.schema.yaml (hunk 5 above)
Comment (PhillypHenning) [reply scripts/helm/bitops.schema.yaml:18]: Fixed.

File: scripts/helm/bitops.schema.yaml (hunk 6 above)
Comment (PhillypHenning) [reply scripts/helm/bitops.schema.yaml:30]: Fixed.

File: scripts/helm/bitops.schema.yaml (hunk 7 above)
Comment (PhillypHenning) [reply scripts/helm/bitops.schema.yaml:37]: Helm was used for testing, I will switch this over to be in the Terraform schema.

**Comment by PhillypHenning** [comment]
New output from Terraform schema + config;
script_options: [  -backend-config="KEY1=foo" -backend-config="KEY2=bar"]

**Comment by PhillypHenning** [comment]
Tested changes by mounting local BitOps dir with changes to a local POC JS app that was deployed to AWS, tested the services (with success) and tore down the infrastructure when complete. 

JS App Service was available before destruction. 

Above changes seem to be working as intended.

File: scripts/bitops-config/README.md [hunk 9]
```diff
@@ -47,6 +47,12 @@ nested:
   foo: true
 ```
 
+### Sample local scripts
+```
+export DEBUG=true; export DEEP_DEBUG=; export BITOPS_DIR=/Users/Phil/Desktop/BitOps/bitops; export SCRIPTS_DIR=$BITOPS_DIR/scripts; export SCHEMA_FILE=$SCRIPTS_DIR/helm/bitops.schema.yaml; export BITOPS_CONFIG_FILE=$BITOPS_DIR/bitops.config.yaml; /Users/Phil/Desktop/BitOps/bitops/scripts/bitops-config/convert-schema.sh $SCHEMA_FILE $BITOPS_CONFIG_FILE
```
Comment (mickmcgrath13) [inline scripts/bitops-config/README.md:52]: maybe remove `/Users/Phil/Desktop/Bitops` and do something like `/path/to/bitops/scripts/bitops-config/convert-schema` ?

even better would be to show how it works from a docker command like:
```
docker run --rm --name bitops-local \
  -e ENVIRONMENT="foo" \
  -e DEBUG="true" \
  -e DEEP_DEBUG="" \
  --entrypoint="/opt/bitops/scripts/bitops-config/convert-schema.sh" \
  -v /path/to/operations:/opt/bitops_deployment \
  bitops:latest \
  "/opt/bitops/scripts/example-tool/bitops.schema.yaml" \
  "/opt/bitops_deployment/foo/example-tool/bitops.config.yaml" 
```

File: scripts/bitops-config/convert-schema.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mickmcgrath13) [reply scripts/bitops-config/convert-schema.sh:69]: @PhillypHenning still need this one :)

**Review by mickmcgrath13 — CHANGES_REQUESTED** [review]

**Comment by mickmcgrath13** [comment]
also, should add the `backend-config` option to the docs here:
https://github.com/bitovi/bitops/blob/master/docs/tool-configuration/configuration-terraform.md

File: scripts/bitops-config/README.md (hunk 9 above)
Comment (PhillypHenning) [reply scripts/bitops-config/README.md:52]: Got it. I remember why I wrote this in the first place. 
So, right under this there is a great example using the same docker container method you use here to test. 

I added the local sample (clumsily it appears) just to point out you could run it as a standalone script.

**Comment by PhillypHenning** [comment]
# ------------------------------------------------------------------------------------------------------------------------------------ #
[Command using double dash optional args]
    philh@Phillyps-MacBook-Pro bitops % export ENV_FILE=.env; \
    export DEBUG=true; \
    export DEEP_DEBUG=true; \
    export BITOPS_DIR=$(pwd); \
    export SCRIPTS_DIR=$BITOPS_DIR/scripts; export SCHEMA_FILE=$SCRIPTS_DIR/terraform/bitops.schema.yaml; \
    export BITOPS_CONFIG_FILE=$BITOPS_DIR/docs/example-config-files/terraform.bitops.config.yaml; \
    $SCRIPTS_DIR/bitops-config/convert-schema.sh $SCHEMA_FILE $BITOPS_CONFIG_FILE --root-key foo --root-key-schema bar

[Output]
    BITOPS SET TO: [/Users/philh/Documents/bitops]
    SCRIPTS SET TO: [/Users/philh/Documents/bitops/scripts]
    SCHEMA_FILESET TO: [/Users/philh/Documents/bitops/scripts/terraform/bitops.schema.yaml]
    BITOPS_CONFIG_FILE SET TO: [/Users/philh/Documents/bitops/docs/example-config-files/terraform.bitops.config.yaml]
    ROOT_KEY SET TO: [foo]
    ROOT_KEY_SCHEMA SET TO: [bar]

# ------------------------------------------------------------------------------------------------------------------------------------ #

[Command using single dash optional args]
    philh@Phillyps-MacBook-Pro bitops % export ENV_FILE=.env; \
    export DEBUG=true; \
    export DEEP_DEBUG=true; \
    export BITOPS_DIR=$(pwd); \
    export SCRIPTS_DIR=$BITOPS_DIR/scripts; export SCHEMA_FILE=$SCRIPTS_DIR/terraform/bitops.schema.yaml; \
    export BITOPS_CONFIG_FILE=$BITOPS_DIR/docs/example-config-files/terraform.bitops.config.yaml; \
    $SCRIPTS_DIR/bitops-config/convert-schema.sh $SCHEMA_FILE $BITOPS_CONFIG_FILE -r foo -rs bar 


[Output]
    BITOPS SET TO: [/Users/philh/Documents/bitops]
    SCRIPTS SET TO: [/Users/philh/Documents/bitops/scripts]
    SCHEMA_FILESET TO: [/Users/philh/Documents/bitops/scripts/terraform/bitops.schema.yaml]
    BITOPS_CONFIG_FILE SET TO: [/Users/philh/Documents/bitops/docs/example-config-files/terraform.bitops.config.yaml]
    ROOT_KEY SET TO: [foo]
    ROOT_KEY_SCHEMA SET TO: [bar]

# ------------------------------------------------------------------------------------------------------------------------------------ #


[Command using positional args]
    philh@Phillyps-MacBook-Pro bitops % export ENV_FILE=.env; \
    export DEBUG=true; \
    export DEEP_DEBUG=true; \
    export BITOPS_DIR=$(pwd); \
    export SCRIPTS_DIR=$BITOPS_DIR/scripts; export SCHEMA_FILE=$SCRIPTS_DIR/terraform/bitops.schema.yaml; \
    export BITOPS_CONFIG_FILE=$BITOPS_DIR/docs/example-config-files/terraform.bitops.config.yaml; \
    $SCRIPTS_DIR/bitops-config/convert-schema.sh $SCHEMA_FILE $BITOPS_CONFIG_FILE foo bar       

[Output]
    BITOPS SET TO: [/Users/philh/Documents/bitops]
    SCRIPTS SET TO: [/Users/philh/Documents/bitops/scripts]
    SCHEMA_FILESET TO: [/Users/philh/Documents/bitops/scripts/terraform/bitops.schema.yaml]
    BITOPS_CONFIG_FILE SET TO: [/Users/philh/Documents/bitops/docs/example-config-files/terraform.bitops.config.yaml]
    ROOT_KEY SET TO: [foo]
    ROOT_KEY_SCHEMA SET TO: [bar]

# ------------------------------------------------------------------------------------------------------------------------------------ #

[Command using no args]
    philh@Phillyps-MacBook-Pro bitops % export ENV_FILE=.env; \
    export DEBUG=true; \
    export DEEP_DEBUG=true; \
    export BITOPS_DIR=$(pwd); \
    export SCRIPTS_DIR=$BITOPS_DIR/scripts; export SCHEMA_FILE=$SCRIPTS_DIR/terraform/bitops.schema.yaml; \
    export BITOPS_CONFIG_FILE=$BITOPS_DIR/docs/example-config-files/terraform.bitops.config.yaml; \
    $SCRIPTS_DIR/bitops-config/convert-schema.sh $SCHEMA_FILE $BITOPS_CONFIG_FILE     

[Output]
    BITOPS SET TO: [/Users/philh/Documents/bitops]
    SCRIPTS SET TO: [/Users/philh/Documents/bitops/scripts]
    SCHEMA_FILESET TO: [/Users/philh/Documents/bitops/scripts/terraform/bitops.schema.yaml]
    BITOPS_CONFIG_FILE SET TO: [/Users/philh/Documents/bitops/docs/example-config-files/terraform.bitops.config.yaml]
    ROOT_KEY SET TO: []
    ROOT_KEY_SCHEMA SET TO: []

# ------------------------------------------------------------------------------------------------------------------------------------ #

**Review by PhillypHenning — COMMENTED** [review]
Making things more complicated then they need to be since 1992

**Comment by PhillypHenning** [comment]
With changes, the output is still the same;
ROOT_KEY SET TO: [foo]
ROOT_KEY_SCHEMA SET TO: [bar]

File: NOTES.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# $BITOPS_SCRIPT_DIR/bitops-config/convert-schema.sh $BITOPS_CONFIG_SCHEMA
+
+export BITOPS_CONFIG_FILE=/Users/philh/Documents/ops_dir/test/terraform/bitops.config.yaml;
+export BITOPS_DIR=$BITOPS_HOME; 
+export BITOPS_CONFIG_SCHEMA=$BITOPS_HOME/scripts/helm/bitops.schema.yaml; 
+export SCRIPTS_DIR=$BITOPS_HOME/scripts; 
+$BITOPS_HOME/scripts/bitops-config/convert-schema.sh $BITOPS_CONFIG_SCHEMA $BITOPS_CONFIG_FILE
+
+
+export ENV_FILE=$BITOPS_HOME/check-me-out-girl.txt;
```
Comment (mickmcgrath13) [inline NOTES.md:10]: 😂

**Review by mickmcgrath13 — APPROVED** [review]

File: NOTES.md (hunk 10 above)
Comment (PhillypHenning) [reply NOTES.md:10]: Oh man that's embarrassing :P

**Pull request merged**